File: FlutterPlugin/android/src/main/kotlin/so/daro/flutter/DaroFlutterPlugin.kt
package so.daro.flutter

import android.app.Application
import android.content.Context
import android.graphics.Color
import android.os.Build
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import droom.daro.SDKConfig
import droom.daro.a.Daro
import droom.daro.core.model.DaroNativeAdChoicePlacement
import so.daro.flutter.native.DaroNativeAdFactory
import so.daro.flutter.native.DaroNativeAdViewFactory
import so.daro.flutter.native.DaroNativeAdPlatformView
import so.daro.flutter.banner.DaroBannerAdViewFactory
import so.daro.flutter.banner.DaroBannerAdPlatformView
import so.daro.flutter.linenative.DaroLineNativeAdViewFactory
import so.daro.flutter.linenative.DaroLineNativeAdPlatformView

class DaroFlutterPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private var adManagerChannel: MethodChannel? = null
  private var interstitialAdManager: DaroInterstitialAdManager? = null
  private var rewardedAdManager: DaroRewardedAdManager? = null
  private var appOpenAdManager: DaroFlutterAppOpenAdManager? = null
  private var lightPopupAdManager: DaroFlutterLightPopupAdManager? = null

  companion object {
    private var nativeAdViewFactory: DaroNativeAdViewFactory? = null

    @JvmStatic
    fun registerNativeAdFactory(factory: DaroNativeAdFactory, factoryId: String) {
      nativeAdViewFactory?.registerFactory(factory, factoryId)
    }
  }

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext

    val interstitialMethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/interstitial")
    val interstitialEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/interstitial_events")

    val manager = DaroInterstitialAdManager()
    interstitialAdManager = manager

    interstitialMethodChannel.setMethodCallHandler(manager)
    interstitialEventChannel.setStreamHandler(manager)

    val rewardedMethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/rewarded")
    val rewardedEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/rewarded_events")

    val rewardedManager = DaroRewardedAdManager()
    rewardedAdManager = rewardedManager

    rewardedMethodChannel.setMethodCallHandler(rewardedManager)
    rewardedEventChannel.setStreamHandler(rewardedManager)

    val appOpenMethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/appopen")
    val appOpenEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/appopen_events")

    val appOpenManager = DaroFlutterAppOpenAdManager()
    appOpenManager.setApplication(context.applicationContext as? Application)
    appOpenAdManager = appOpenManager

    appOpenMethodChannel.setMethodCallHandler(appOpenManager)
    appOpenEventChannel.setStreamHandler(appOpenManager)

    val lightPopupMethodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/lightpopup")
    val lightPopupEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/lightpopup_events")

    val lightPopupManager = DaroFlutterLightPopupAdManager()
    lightPopupManager.setContext(context)
    lightPopupAdManager = lightPopupManager

    lightPopupMethodChannel.setMethodCallHandler(lightPopupManager)
    lightPopupEventChannel.setStreamHandler(lightPopupManager)

    // ad_manager 채널: view-based 광고 로드/파괴
    val adMgrChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "daro_flutter/ad_manager")
    adManagerChannel = adMgrChannel
    adMgrChannel.setMethodCallHandler { call, result ->
      handleAdManager(call, result, adMgrChannel)
    }

    val nativeFactory = DaroNativeAdViewFactory(context, flutterPluginBinding.binaryMessenger)
    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "daro_native_ad_view",
      nativeFactory
    )
    nativeAdViewFactory = nativeFactory

    val bannerFactory = DaroBannerAdViewFactory(context, flutterPluginBinding.binaryMessenger)
    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "daro_banner_ad_view",
      bannerFactory
    )

    val lineNativeFactory = DaroLineNativeAdViewFactory(context, flutterPluginBinding.binaryMessenger)
    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "daro_line_native_ad_view",
      lineNativeFactory
    )
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${Build.VERSION.RELEASE}")
      }
      "initialize" -> {
        initializeDaroSDK(call, result)
      }
      "setAppMuted" -> {
        val muted = call.arguments as? Boolean ?: false
        Daro.setAppMuted(muted)
        result.success(null)
      }
      "setUserId" -> {
        val userId = call.arguments as? String ?: ""
        Daro.setUserId(context, userId)
        result.success(null)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun handleAdManager(call: MethodCall, result: Result, channel: MethodChannel) {
    val args = call.arguments as? Map<*, *> ?: emptyMap<String, Any>()

    when (call.method) {
      "loadBannerAd" -> {
        val adId = args["adId"] as? Int
            ?: return result.error("INVALID_ARGS", "Missing adId", null)
        val adUnitId = args["adUnitId"] as? String
            ?: return result.error("INVALID_ARGS", "Missing adUnitId", null)
        val sizeString = args["size"] as? String
            ?: return result.error("INVALID_ARGS", "Missing size", null)

        val platformView = DaroBannerAdPlatformView(
          context, adId, adUnitId, sizeString, channel
        )
        DaroAdInstanceManager.store(platformView, adId)
        result.success(null)
      }

      "loadNativeAd" -> {
        val adId = args["adId"] as? Int
            ?: return result.error("INVALID_ARGS", "Missing adId", null)
        val adUnitId = args["adUnitId"] as? String
            ?: return result.error("INVALID_ARGS", "Missing adUnitId", null)
        val factoryId = args["factoryId"] as? String
            ?: return result.error("INVALID_ARGS", "Missing factoryId", null)

        val factory = nativeAdViewFactory?.getFactory(factoryId)
            ?: return result.error("FACTORY_NOT_FOUND", "Factory not registered: $factoryId", null)

        val positionName = args["preferredAdChoicesPosition"] as? String
        val placement = positionName.toDaroNativeAdChoicePlacement()

        val platformView = DaroNativeAdPlatformView(
          context, adId, adUnitId, factory, placement, channel
        )
        DaroAdInstanceManager.store(platformView, adId)
        result.success(null)
      }

      "loadLineNativeAd" -> {
        val adId = args["adId"] as? Int
            ?: return result.error("INVALID_ARGS", "Missing adId", null)
        val adUnitId = args["adUnitId"] as? String
            ?: return result.error("INVALID_ARGS", "Missing adUnitId", null)

        val backgroundColor = parseColorFromMap(args["backgroundColor"])
        val contentColor = parseColorFromMap(args["contentColor"])
        val adMarkLabelTextColor = parseColorFromMap(args["adMarkLabelTextColor"])
        val adMarkLabelBackgroundColor = parseColorFromMap(args["adMarkLabelBackgroundColor"])

        val platformView = DaroLineNativeAdPlatformView(
          context, adId, adUnitId,
          backgroundColor, contentColor,
          adMarkLabelTextColor, adMarkLabelBackgroundColor,
          channel
        )
        DaroAdInstanceManager.store(platformView, adId)
        result.success(null)
      }

      "disposeAd" -> {
        val adId = args["adId"] as? Int
            ?: return result.error("INVALID_ARGS", "Missing adId", null)
        DaroAdInstanceManager.remove(adId)
        result.success(null)
      }

      else -> result.notImplemented()
    }
  }

  private fun parseColorFromMap(value: Any?): Int? {
    val map = value as? Map<*, *> ?: return null
    val r = (map["r"] as? Number)?.toInt() ?: return null
    val g = (map["g"] as? Number)?.toInt() ?: return null
    val b = (map["b"] as? Number)?.toInt() ?: return null
    val a = (map["a"] as? Number)?.toInt() ?: return null
    return Color.argb(a, r, g, b)
  }

  private fun initializeDaroSDK(call: MethodCall, result: Result) {
    try {
      val application = context.applicationContext as? Application
      if (application == null) {
        result.error(
          "INIT_ERROR",
          "Failed to get Application context",
          null
        )
        return
      }

      val args = call.arguments as? Map<*, *>
      val isDebugMode = args?.get("isDebugMode") as? Boolean ?: false

      Daro.init(
        application = application,
        sdkConfig = SDKConfig.Builder()
          .setDebugMode(isDebugMode)
          .build()
      )

      val hasGdprConsent = args?.get("hasGdprConsent") as? Boolean
      val gdprString = args?.get("gdprConsentString") as? String ?: ""
      hasGdprConsent?.let {
        if (it) Daro.grantGdpr(context, gdprString)
        else Daro.rejectGdpr(context, gdprString)
      }

      val doNotSell = args?.get("doNotSell") as? Boolean
      val ccpaString = args?.get("ccpaConsentString") as? String ?: ""
      doNotSell?.let {
        if (it) Daro.rejectCCPA(context, ccpaString)
        else Daro.grantCCPA(context, ccpaString)
      }

      val coppa = args?.get("isTaggedForChildDirectedTreatment") as? Boolean
      coppa?.let {
        if (it) Daro.grantCOPPA(context)
        else Daro.rejectCOPPA(context)
      }

      result.success(null)
    } catch (e: Exception) {
      result.error(
        "INIT_ERROR",
        "Daro SDK initialization failed: ${e.message}",
        e.toString()
      )
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    adManagerChannel?.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    interstitialAdManager?.setActivity(binding.activity)
    rewardedAdManager?.setActivity(binding.activity)
    appOpenAdManager?.setActivity(binding.activity)
    lightPopupAdManager?.setActivity(binding.activity)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    interstitialAdManager?.setActivity(null)
    rewardedAdManager?.setActivity(null)
    appOpenAdManager?.setActivity(null)
    lightPopupAdManager?.setActivity(null)
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    interstitialAdManager?.setActivity(binding.activity)
    rewardedAdManager?.setActivity(binding.activity)
    appOpenAdManager?.setActivity(binding.activity)
    lightPopupAdManager?.setActivity(binding.activity)
  }

  override fun onDetachedFromActivity() {
    interstitialAdManager?.setActivity(null)
    rewardedAdManager?.setActivity(null)
    appOpenAdManager?.setActivity(null)
    lightPopupAdManager?.setActivity(null)
  }
}

private fun String?.toDaroNativeAdChoicePlacement(): DaroNativeAdChoicePlacement = when (this) {
    "topLeft"     -> DaroNativeAdChoicePlacement.TOP_LEFT
    "topRight"    -> DaroNativeAdChoicePlacement.TOP_RIGHT
    "bottomLeft"  -> DaroNativeAdChoicePlacement.BOTTOM_LEFT
    "bottomRight" -> DaroNativeAdChoicePlacement.BOTTOM_RIGHT
    else          -> DaroNativeAdChoicePlacement.BOTTOM_RIGHT
}

File: FlutterPlugin/ios/Classes/DaroFlutterPlugin.swift
import Flutter
import UIKit
import Daro

public class DaroFlutterPlugin: NSObject, FlutterPlugin {
  private var interstitialAdManager: DaroInterstitialAdManager?
  private var rewardedAdManager: DaroRewardedAdManager?
  private var appOpenAdManager: DaroAppOpenAdManager?
  private var lightPopupAdManager: DaroLightPopupAdManager?
  private static var nativeAdViewFactory: DaroNativeAdViewFactory?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "daro_flutter", binaryMessenger: registrar.messenger())
    let instance = DaroFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)

    let interstitialMethodChannel = FlutterMethodChannel(name: "daro_flutter/interstitial", binaryMessenger: registrar.messenger())
    let interstitialEventChannel = FlutterEventChannel(name: "daro_flutter/interstitial_events", binaryMessenger: registrar.messenger())

    let interstitialManager = DaroInterstitialAdManager()
    instance.interstitialAdManager = interstitialManager

    interstitialMethodChannel.setMethodCallHandler { call, result in
      interstitialManager.handle(call, result: result)
    }
    interstitialEventChannel.setStreamHandler(interstitialManager)

    let rewardedMethodChannel = FlutterMethodChannel(name: "daro_flutter/rewarded", binaryMessenger: registrar.messenger())
    let rewardedEventChannel = FlutterEventChannel(name: "daro_flutter/rewarded_events", binaryMessenger: registrar.messenger())

    let rewardedManager = DaroRewardedAdManager()
    instance.rewardedAdManager = rewardedManager

    rewardedMethodChannel.setMethodCallHandler { call, result in
      rewardedManager.handle(call, result: result)
    }
    rewardedEventChannel.setStreamHandler(rewardedManager)

    let appOpenMethodChannel = FlutterMethodChannel(name: "daro_flutter/appopen", binaryMessenger: registrar.messenger())
    let appOpenEventChannel = FlutterEventChannel(name: "daro_flutter/appopen_events", binaryMessenger: registrar.messenger())

    let appOpenManager = DaroAppOpenAdManager()
    instance.appOpenAdManager = appOpenManager

    appOpenMethodChannel.setMethodCallHandler { call, result in
      appOpenManager.handle(call, result: result)
    }
    appOpenEventChannel.setStreamHandler(appOpenManager)

    let lightPopupMethodChannel = FlutterMethodChannel(name: "daro_flutter/lightpopup", binaryMessenger: registrar.messenger())
    let lightPopupEventChannel = FlutterEventChannel(name: "daro_flutter/lightpopup_events", binaryMessenger: registrar.messenger())

    let lightPopupManager = DaroLightPopupAdManager()
    instance.lightPopupAdManager = lightPopupManager

    lightPopupMethodChannel.setMethodCallHandler { call, result in
      lightPopupManager.handle(call, result: result)
    }
    lightPopupEventChannel.setStreamHandler(lightPopupManager)

    // ad_manager 채널: view-based 광고(배너/네이티브/라인네이티브) 로드/파괴
    let adManagerChannel = FlutterMethodChannel(
      name: "daro_flutter/ad_manager",
      binaryMessenger: registrar.messenger()
    )
    adManagerChannel.setMethodCallHandler { call, result in
      instance.handleAdManager(call, result: result, messenger: registrar.messenger(), channel: adManagerChannel)
    }

    // 네이티브 광고 팩토리 등록
    let nativeFactory = DaroNativeAdViewFactory(messenger: registrar.messenger())
    registrar.register(
      nativeFactory,
      withId: "daro_native_ad_view"
    )
    Self.nativeAdViewFactory = nativeFactory

    // 배너 광고 팩토리 등록
    let bannerFactory = DaroBannerAdViewFactory(messenger: registrar.messenger())
    registrar.register(
      bannerFactory,
      withId: "daro_banner_ad_view"
    )

    // 라인 네이티브 광고 팩토리 등록
    let lineNativeFactory = DaroLineNativeAdViewFactory(messenger: registrar.messenger())
    registrar.register(
      lineNativeFactory,
      withId: "daro_line_native_ad_view"
    )
  }

  /// 네이티브 광고 팩토리를 등록할 수 있는 공개 메서드
  public static func registerNativeAdFactory(
    _ factory: DaroNativeAdFactory,
    factoryId: String
  ) {
    nativeAdViewFactory?.registerFactory(factory, withId: factoryId)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "initialize":
      initializeDaroSDK(call: call, result: result)
    case "setAppMuted":
      let muted = call.arguments as? Bool ?? false
      DaroAds.shared.setAppMuted(muted)
      result(nil)
    case "setUserId":
      let userId = call.arguments as? String
      DaroAds.shared.userId = userId
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleAdManager(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult,
    messenger: FlutterBinaryMessenger,
    channel: FlutterMethodChannel
  ) {
    let args = call.arguments as? [String: Any] ?? [:]

    switch call.method {
    case "loadBannerAd":
      guard let adId = args["adId"] as? Int,
            let adUnitId = args["adUnitId"] as? String,
            let sizeString = args["size"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
        return
      }

      let bannerSize: DaroAdBannerSize = sizeString == "mrec" ? .MREC : .banner
      let platformView = DaroBannerAdPlatformView(
        frame: .zero,
        adId: adId,
        adUnitId: adUnitId,
        bannerSize: bannerSize,
        channel: channel
      )
      DaroAdInstanceManager.shared.store(platformView, forId: adId)
      result(nil)

    case "loadNativeAd":
      guard let adId = args["adId"] as? Int,
            let adUnitId = args["adUnitId"] as? String,
            let factoryId = args["factoryId"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
        return
      }

      guard let factory = Self.nativeAdViewFactory?.getFactory(withId: factoryId) else {
        result(FlutterError(code: "FACTORY_NOT_FOUND", message: "Factory not registered: \(factoryId)", details: nil))
        return
      }

      let platformView = DaroNativeAdPlatformView(
        frame: .zero,
        adId: adId,
        adUnitId: adUnitId,
        factory: factory,
        channel: channel
      )
      DaroAdInstanceManager.shared.store(platformView, forId: adId)
      result(nil)

    case "loadLineNativeAd":
      guard let adId = args["adId"] as? Int,
            let adUnitId = args["adUnitId"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
        return
      }

      let configuration = DaroLineNativeAdConfiguration()

      if let backgroundColor = args["backgroundColor"] as? [String: Any],
         let color = colorFromMap(backgroundColor) {
        configuration.backgroundColor = color
      }
      if let contentColor = args["contentColor"] as? [String: Any],
         let color = colorFromMap(contentColor) {
        configuration.titleTextColor = color
      }
      if let adMarkLabelTextColor = args["adMarkLabelTextColor"] as? [String: Any],
         let color = colorFromMap(adMarkLabelTextColor) {
        configuration.adMarkTextColor = color
      }
      if let adMarkLabelBackgroundColor = args["adMarkLabelBackgroundColor"] as? [String: Any],
         let color = colorFromMap(adMarkLabelBackgroundColor) {
        configuration.adMarkBackgroundColor = color
      }

      let platformView = DaroLineNativeAdPlatformView(
        frame: .zero,
        adId: adId,
        adUnitId: adUnitId,
        configuration: configuration,
        channel: channel
      )
      DaroAdInstanceManager.shared.store(platformView, forId: adId)
      result(nil)

    case "disposeAd":
      guard let adId = args["adId"] as? Int else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing adId", details: nil))
        return
      }
      DaroAdInstanceManager.shared.remove(forId: adId)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func colorFromMap(_ map: [String: Any]) -> UIColor? {
    guard let r = map["r"] as? Int,
          let g = map["g"] as? Int,
          let b = map["b"] as? Int,
          let a = map["a"] as? Int else {
      return nil
    }
    return UIColor(
      red: CGFloat(r) / 255.0,
      green: CGFloat(g) / 255.0,
      blue: CGFloat(b) / 255.0,
      alpha: CGFloat(a) / 255.0
    )
  }

  private func initializeDaroSDK(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]

    if let isDebug = args?["isDebugMode"] as? Bool {
      DaroAds.shared.logLevel = isDebug ? .debug : .off
    }
    if let hasConsent = args?["hasGdprConsent"] as? Bool {
      DaroAds.shared.hasUserConsent = hasConsent
    }
    if let gdprString = args?["gdprConsentString"] as? String {
      DaroAds.shared.gdprConsentString = gdprString
    }
    if let doNotSell = args?["doNotSell"] as? Bool {
      DaroAds.shared.doNotSell = doNotSell
    }
    if let ccpaString = args?["ccpaConsentString"] as? String {
      DaroAds.shared.ccpaString = ccpaString
    }
    if let coppa = args?["isTaggedForChildDirectedTreatment"] as? Bool {
      DaroAds.shared.isTaggedForChildDirectedTreatment = coppa
    }

    DaroAds.shared.initialized { error in
      if let error = error {
        NSLog("[DaroFlutter] initialization failed: \(error.localizedDescription)")
      }
    }
    result(nil)
  }
}
